<a href="https://colab.research.google.com/github/MicuPaul/Datasets_join/blob/main/Datasets_join.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Importing packages**

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
!unzip /content/drive/MyDrive/datasets.zip

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Archive:  /content/drive/MyDrive/datasets.zip
  inflating: website_dataset.csv     
  inflating: __MACOSX/._website_dataset.csv  
  inflating: facebook_dataset.csv    
  inflating: __MACOSX/._facebook_dataset.csv  
  inflating: google_dataset.csv      
  inflating: __MACOSX/._google_dataset.csv  


#**Opening the datasets**

In [2]:
facebook_dt = pd.read_csv("/content/facebook_dataset.csv", on_bad_lines='skip')
google_dt = pd.read_csv("/content/google_dataset.csv", on_bad_lines='skip', low_memory=False)
website_dt = pd.read_csv("/content/website_dataset.csv", sep = ";" ,on_bad_lines='skip') #Website data is separated with a semi colon

The on_bad_lines = "skip" lets us skip lines where there are more data fields than the expected amount.

#**Visualizing the data**

In [3]:
google_dt.columns.values

array(['address', 'category', 'city', 'country_code', 'country_name',
       'name', 'phone', 'phone_country_code', 'raw_address', 'raw_phone',
       'region_code', 'region_name', 'text', 'zip_code', 'domain'],
      dtype=object)

In [4]:
google_dt.head()

address  \
0  28 Central Coast Hwy, West Gosford NSW 2250, A...   
1   400 Scott St, St. Catharines, ON L2M 3W2, Canada   
2      191 Pleasant St, Yarmouth, NS B5A 2J9, Canada   
3  11040 Santa Monica Blvd Suite 370, Los Angeles...   
4                                                NaN   

                            category            city country_code  \
0            Fabric-Based Home Goods         gosford           au   
1                        Book Stores  st. catharines           ca   
2  Other Building Material Retailers        yarmouth           ca   
3            Plastic Surgery Clinics     los angeles           us   
4                Catering & Delivery             NaN          NaN   

    country_name                                name         phone  \
0      australia              Spotlight West Gosford  +61243355946   
1         canada       Heritage Christian Book Store  +19059374553   
2         canada                Pleasant Timber Mart  +19027429181   
3  united states  Skin Specifics Medical Spa West LA  +18184268353   
4            NaN                       Lakay express           NaN   

  phone_country_code                                        raw_address  \
0                 au                        West Gosford NSW, Australia   
1                 ca                   400 Scott St · In Grantham Plaza   
2                 ca        7+ years in business · Yarmouth, NS, Canada   
3                 us  7+ years in business · 11040 Santa Monica Blvd...   
4                NaN                                                NaN   

         raw_phone region_code      region_name  \
0  +61 2 4335 5946         nsw  new south wales   
1  +1 905-937-4553          on          ontario   
2  +1 902-742-9181          ns      nova scotia   
3  +1 818-426-8353          ca       california   
4              NaN         NaN              NaN   

                                                text zip_code  \
0  4.1 (766) · Craft store West Gosford NSW, Aust...     2250   
1  4.7 (100) · Book store 400 Scott St · In Grant...  l2m 3w2   
2  4.7 (40) · Building materials store 7+ years i...  b5a 2j9   
3  4.3 (15) · Medical spa 7+ years in business · ...    90025   
4         5.0 (1) · Mobile caterer Open ⋅ Closes 9PM      NaN   

                domain  
0  spotlightstores.com  
1      bookmanager.com  
2        timbermart.ca  
3            linktr.ee  
4            linktr.ee

In [5]:
facebook_dt.columns.values

array(['domain', 'address', 'categories', 'city', 'country_code',
       'country_name', 'description', 'email', 'link', 'name',
       'page_type', 'phone', 'phone_country_code', 'region_code',
       'region_name', 'zip_code'], dtype=object)

In [6]:
facebook_dt.head()

domain                                            address  \
0       euro-hygiene-34.fr  134 rue entrepreneurs, za du vigné, 30420, cal...   
1  lakesidehomeservices.ca                                                NaN   
2        rossiterboats.com                                                NaN   
3     caa-architecture.com  601 cordova st w # 270, v6b 1g1, vancouver, bc...   
4            apexsurety.ca  unit 3 - 4 donald street, r3l 2t7, winnipeg, m...   

                                          categories       city country_code  \
0                                                NaN  calvisson           fr   
1  Appliance Repair & Maintenance|Home Builders &...        NaN           ca   
2        Boats & Yachts Dealers|Boat Tours & Cruises        NaN          NaN   
3  Architects & Architectural Services|Other Engi...  vancouver           ca   
4                                                NaN   winnipeg           ca   

  country_name                                        description  \
0       france                                                NaN   
1          NaN                                                NaN   
2          NaN  At Rossiter Boats we build semi-custom power b...   
3       canada                                                NaN   
4       canada                                                NaN   

                        email                             link  \
0                         NaN       https://euro-hygiene-34.fr   
1                         NaN  https://lakesidehomeservices.ca   
2      info@rossiterboats.com         http://rossiterboats.com   
3  office@caa-arcitecture.com      http://caa-architecture.com   
4                         NaN            https://apexsurety.ca   

                                    name      page_type         phone  \
0                           Euro Hygiène  LocalBusiness           NaN   
1                 Lakeside Home Services  LocalBusiness  1.613832e+10   
2                         Rossiter Boats  LocalBusiness           NaN   
3  Chandler Associates Architecture Inc.  LocalBusiness  1.604687e+10   
4           Apex Surety & Insurance Ltd.  LocalBusiness  1.204560e+10   

  phone_country_code region_code       region_name zip_code  
0                NaN         occ         occitanie    30420  
1                 ca         NaN               NaN      NaN  
2                NaN         NaN               NaN      NaN  
3                 ca          bc  british columbia  v6b 1g1  
4                 ca          mb          manitoba  r3l 2t7

In [7]:
website_dt.columns.values

array(['root_domain', 'domain_suffix', 'language', 'legal_name',
       'main_city', 'main_country', 'main_region', 'phone', 'site_name',
       'tld', 's_category'], dtype=object)

In [8]:
website_dt.head()

root_domain domain_suffix language  \
0     converterman.com           NaN      NaN   
1  clothesencounter.ca            ca       en   
2       investa.com.au        com.au       en   
3    timminsgarage.com           com       en   
4         destinate.ca            ca       en   

                                   legal_name main_city main_country  \
0                                         NaN       NaN          NaN   
1                                         NaN  cardigan       canada   
2  Investa Wholesale Funds Management Limited  brisbane    australia   
3                         Timmins Garage Inc.   timmins       canada   
4                        Destinate Group Ltd.       NaN       Canada   

            main_region        phone               site_name  tld  \
0                   NaN          NaN                     NaN  NaN   
1  prince edward island  13066937766       Clothes Encounter   ca   
2            queensland  61282269300  Investa Property Group   au   
3               ontario  18775896640          Timmins Garage  com   
4                   NaN          NaN         Destinate Group   ca   

                           s_category  
0                                 NaN  
1       Shoes & Other Footwear Stores  
2              Real Estate Developers  
3  Automobile Dealers & Manufacturers  
4                 Business Consulting

#**Handling missing values**

First we will clean the datasets by handling the missing values. If there are missing values in the Category, Address (country, region...), Phone, Company name fields, we will drop that entry.
I'll start with the google dataset.

In [9]:
old_size = google_dt.shape
google_dt.dropna(subset=['address', 'category', 'country_code', 'city', 'name', 'phone', 'region_code', 'domain', 'zip_code'], inplace=True)
new_size = google_dt.shape
print("Number of Rows(data): ", new_size[0] , "Rows dropped: " ,  old_size[0] - new_size[0]) #Shows the number of entries droped

Number of Rows(data):  195757 Rows dropped:  151168


The method above (dropna) checks the columns specified and drops the row if there is a missing value in that column. The inplace specifies if the new dataframe is a copy or the original is modified.

Next we do the same for the other two datasets.

In [10]:
old_size = facebook_dt.shape
facebook_dt.dropna(subset=['address', 'categories', 'country_code', 'city', 'name', 'phone', 'region_code', 'domain', 'zip_code'], inplace=True)
new_size = facebook_dt.shape
print("Number of Rows(data): ", new_size[0] , "Rows dropped: " ,  old_size[0] - new_size[0]) #Shows the number of entries droped

Number of Rows(data):  19496 Rows dropped:  51671


In [11]:
old_size = website_dt.shape
website_dt.dropna(subset=['main_city', 'main_country', 'main_region', 'phone', 'legal_name', 'root_domain', 's_category'], inplace=True)
new_size = website_dt.shape
print("Number of Rows(data): ", new_size[0] , "Rows dropped: " ,  old_size[0] - new_size[0]) #Shows the number of entries droped

Number of Rows(data):  27944 Rows dropped:  44074


#**Joining the datasets**

I've decided to use the company name the merge key do to it being mostly unique and also because I have experimented using other merge keys and I have seen that using more merge keys will not find the data that is in both datasets and joins them. For example when using the name as the merge key the merged dataset between the facebook and google datasets has 208609 rows meaning that (195757(rows google) + 19496(rows facebook) - 208609(rows merged)) 6644 data was found in the facebook datasets with the same company name as in the google dataset.

In [12]:
g_and_f = pd.merge(google_dt, facebook_dt,  how='outer',  left_on=['name'], right_on=['name'])
g_and_f.shape

(208609, 30)

In [13]:
display(g_and_f)

address_x  \
0       28 Central Coast Hwy, West Gosford NSW 2250, A...   
1        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
2        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
3           191 Pleasant St, Yarmouth, NS B5A 2J9, Canada   
4       11040 Santa Monica Blvd Suite 370, Los Angeles...   
...                                                   ...   
208604                                                NaN   
208605                                                NaN   
208606                                                NaN   
208607                                                NaN   
208608                                                NaN   

                                 category          city_x country_code_x  \
0                 Fabric-Based Home Goods         gosford             au   
1                             Book Stores  st. catharines             ca   
2                             Book Stores  st. catharines             ca   
3       Other Building Material Retailers        yarmouth             ca   
4                 Plastic Surgery Clinics     los angeles             us   
...                                   ...             ...            ...   
208604                                NaN             NaN            NaN   
208605                                NaN             NaN            NaN   
208606                                NaN             NaN            NaN   
208607                                NaN             NaN            NaN   
208608                                NaN             NaN            NaN   

       country_name_x                                name       phone_x  \
0           australia              Spotlight West Gosford  +61243355946   
1              canada       Heritage Christian Book Store  +19059374553   
2              canada       Heritage Christian Book Store  +19059374553   
3              canada                Pleasant Timber Mart  +19027429181   
4       united states  Skin Specifics Medical Spa West LA  +18184268353   
...               ...                                 ...           ...   
208604            NaN                McLean Auctions Ltd.           NaN   
208605            NaN                  Promenade Printing           NaN   
208606            NaN  Hardin County Water District No. 1           NaN   
208607            NaN       Serenity Skin Care Studio Inc           NaN   
208608            NaN              East Toronto Foot Care           NaN   

       phone_country_code_x  \
0                        au   
1                        ca   
2                        ca   
3                        ca   
4                        us   
...                     ...   
208604                  NaN   
208605                  NaN   
208606                  NaN   
208607                  NaN   
208608                  NaN   

                                              raw_address        raw_phone  \
0                             West Gosford NSW, Australia  +61 2 4335 5946   
1                        400 Scott St · In Grantham Plaza  +1 905-937-4553   
2                        400 Scott St · In Grantham Plaza  +1 905-937-4553   
3             7+ years in business · Yarmouth, NS, Canada  +1 902-742-9181   
4       7+ years in business · 11040 Santa Monica Blvd...  +1 818-426-8353   
...                                                   ...              ...   
208604                                                NaN              NaN   
208605                                                NaN              NaN   
208606                                                NaN              NaN   
208607                                                NaN              NaN   
208608                                                NaN              NaN   

        ... country_name_y                                        description  \
0       ...            NaN                                                NaN   
1       ...            NaN   

In [14]:
g_and_f.columns.values

array(['address_x', 'category', 'city_x', 'country_code_x',
       'country_name_x', 'name', 'phone_x', 'phone_country_code_x',
       'raw_address', 'raw_phone', 'region_code_x', 'region_name_x',
       'text', 'zip_code_x', 'domain_x', 'domain_y', 'address_y',
       'categories', 'city_y', 'country_code_y', 'country_name_y',
       'description', 'email', 'link', 'page_type', 'phone_y',
       'phone_country_code_y', 'region_code_y', 'region_name_y',
       'zip_code_y'], dtype=object)

First I will delete the raw_address and raw_phone columns as they are not needed because we have the normal address and phone columns.

In [15]:
del g_and_f['raw_address']
del g_and_f['raw_phone']

As we can see because we used only the name to join we have a lot of column name conflits because a lot of columns have the same name in both datasets. I have decided to keep the columns from the google dataset because it is overall a bigger dataset and the data is cleaner. For the values that are null in the google columns i will copy the data from the facebook columns. For example if adress_x is null i will copy adress_y to it and then will drop adress_y.
I will do this with all conflicting columns.

In [16]:
#Handling address columns
g_and_f.loc[g_and_f["address_x"].isnull(), "address_x"] = g_and_f["address_y"]
del g_and_f['address_y']
g_and_f.rename(columns = {'address_x':'address'}, inplace = True)

The loc method just replaces all the values that are null in the adress_x column with the value in the same row of column adress_y.
Now we do this with all conflicting columns.

In [17]:
#Handling category columns
g_and_f.loc[g_and_f["category"].isnull(), "category"] = g_and_f["categories"]
del g_and_f['categories']

In [18]:
#Handling city columns
g_and_f.loc[g_and_f["city_x"].isnull(), "city_x"] = g_and_f["city_y"]
del g_and_f['city_y']
g_and_f.rename(columns = {'city_x':'city'}, inplace = True)

In [19]:
#Handling country_code columns
g_and_f.loc[g_and_f["country_code_x"].isnull(), "country_code_x"] = g_and_f["country_code_y"]
del g_and_f['country_code_y']
g_and_f.rename(columns = {'country_code_x':'country_code'}, inplace = True)

In [20]:
#Handling country_name columns
g_and_f.loc[g_and_f["country_name_x"].isnull(), "country_name_x"] = g_and_f["country_name_y"]
del g_and_f['country_name_y']
g_and_f.rename(columns = {'country_name_x':'country_name'}, inplace = True)

In [21]:
#Handling phone columns
g_and_f.loc[g_and_f["phone_x"].isnull(), "phone_x"] = g_and_f["phone_y"]
del g_and_f['phone_y']
g_and_f.rename(columns = {'phone_x':'phone'}, inplace = True)

In [22]:
#Handling phone_country_code columns
g_and_f.loc[g_and_f["phone_country_code_x"].isnull(), "phone_country_code_x"] = g_and_f["phone_country_code_y"]
del g_and_f['phone_country_code_y']
g_and_f.rename(columns = {'phone_country_code_x':'phone_country_code'}, inplace = True)

In [23]:
#Handling region_code columns
g_and_f.loc[g_and_f["region_code_x"].isnull(), "region_code_x"] = g_and_f["region_code_y"]
del g_and_f['region_code_y']
g_and_f.rename(columns = {'region_code_x':'region_code'}, inplace = True)

In [24]:
#Handling region_name columns
g_and_f.loc[g_and_f["region_name_x"].isnull(), "region_name_x"] = g_and_f["region_name_y"]
del g_and_f['region_name_y']
g_and_f.rename(columns = {'region_name_x':'region_name'}, inplace = True)

In [25]:
#Handling description columns
g_and_f.loc[g_and_f["text"].isnull(), "text"] = g_and_f["description"]
del g_and_f['description']
g_and_f.rename(columns = {'text':'description'}, inplace = True)

In [26]:
#Handling zip_code columns
g_and_f.loc[g_and_f["zip_code_x"].isnull(), "zip_code_x"] = g_and_f["zip_code_y"]
del g_and_f['zip_code_y']
g_and_f.rename(columns = {'zip_code_x':'zip_code'}, inplace = True)

In [27]:
#Handling domain columns
g_and_f.loc[g_and_f["domain_x"].isnull(), "domain_x"] = g_and_f["domain_y"]
del g_and_f['domain_y']
g_and_f.rename(columns = {'domain_x':'domain'}, inplace = True)

Now that we are done with the columns i will check wether there are any missing values in the import columns (address, category, country_code, city, name,phone, region_code, domain, zip_code)

In [28]:
print(g_and_f.isnull().any())

address               False
category              False
city                  False
country_code          False
country_name          False
name                  False
phone                 False
phone_country_code     True
region_code           False
region_name           False
description            True
zip_code              False
domain                False
email                  True
link                   True
page_type              True
dtype: bool


Our most important columns dont have any missing values so everything looks good.

In [29]:
display(g_and_f)

address  \
0       28 Central Coast Hwy, West Gosford NSW 2250, A...   
1        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
2        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
3           191 Pleasant St, Yarmouth, NS B5A 2J9, Canada   
4       11040 Santa Monica Blvd Suite 370, Los Angeles...   
...                                                   ...   
208604  2140 little britain road, k9v 4r2, lindsay, on...   
208605  15615 coit road, suite 101, 75248, dallas, tx,...   
208606  1400 rogersville rd, 40160, radcliff, ky, kent...   
208607  15 westmount rd s, unit 301, n2r 1p1, kitchene...   
208608  643 queen street east, m4m 1g4, toronto, on, c...   

                                                 category            city  \
0                                 Fabric-Based Home Goods         gosford   
1                                             Book Stores  st. catharines   
2                                             Book Stores  st. catharines   
3                       Other Building Material Retailers        yarmouth   
4                                 Plastic Surgery Clinics     los angeles   
...                                                   ...             ...   
208604                                      Art Galleries         lindsay   
208605  Digital & Marketing Agencies|Graphic Design|Co...          dallas   
208606  Utility companies|Auto Services|Utility companies        radcliff   
208607        Hair Removal Service|Spa & Wellness Centers       kitchener   
208608  Clinics - Surgeons & Physicians|Clinics - Surg...         toronto   

       country_code   country_name                                name  \
0                au      australia              Spotlight West Gosford   
1                ca         canada       Heritage Christian Book Store   
2                ca         canada       Heritage Christian Book Store   
3                ca         canada                Pleasant Timber Mart   
4                us  united states  Skin Specifics Medical Spa West LA   
...             ...            ...                                 ...   
208604           ca         canada                McLean Auctions Ltd.   
208605           us  united states                  Promenade Printing   
208606           us  united states  Hardin County Water District No. 1   
208607           ca         canada       Serenity Skin Care Studio Inc   
208608           ca         canada              East Toronto Foot Care   

                phone phone_country_code region_code      region_name  \
0        +61243355946                 au         nsw  new south wales   
1        +19059374553                 ca          on          ontario   
2        +19059374553                 ca          on          ontario   
3        +19027429181                 ca          ns      nova scotia   
4        +18184268353                 us          ca       california   
...               ...                ...         ...              ...   
208604  17053242783.0                 ca          on          ontario   
208605  19722340760.0                NaN          tx            texas   
208606  12703513222.0                 us          ky         kentucky   
208607  15197472256.0                 ca          on          ontario   
208608  14167924420.0                 ca          on          ontario   

                                              description zip_code  \
0       4.1 (766) · Craft store West Gosford NSW, Aust...     2250   
1       4.7 (100) · Book store 400 Scott St · In Grant...  l2m 3w2   
2       4.7 (101) · Book store 400 Scott St · In Grant...  l2m 3w2   
3       4.7 (40) · Building materials store 7+ years i...  b5a 2j9   
4       4.3 (15) · Medical spa 7+ years in business · ...    90025   
...                                                   ...      ...   
208604                                                NaN  k9v 4r2   
208605                                                NaN    75248

In [30]:
g_and_f.columns.values

array(['address', 'category', 'city', 'country_code', 'country_name',
       'name', 'phone', 'phone_country_code', 'region_code',
       'region_name', 'description', 'zip_code', 'domain', 'email',
       'link', 'page_type'], dtype=object)

Next we will do the final merge with the g_and_f dataset and the website_dt using the same principle and also using the company name to merge.

(195757(rows google) + 19496(rows facebook) + 27944(rows website) - 232118(rows merged_all))

11079 data rows with the same company name in all three datasets.

In [31]:
Datasets_joined = pd.merge(g_and_f, website_dt,  how='outer',  left_on=['name'], right_on=['legal_name'])
Datasets_joined.shape

(232118, 27)

In [32]:
display(Datasets_joined)

address  \
0       28 Central Coast Hwy, West Gosford NSW 2250, A...   
1        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
2        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
3           191 Pleasant St, Yarmouth, NS B5A 2J9, Canada   
4       11040 Santa Monica Blvd Suite 370, Los Angeles...   
...                                                   ...   
232113                                                NaN   
232114                                                NaN   
232115                                                NaN   
232116                                                NaN   
232117                                                NaN   

                                 category            city country_code  \
0                 Fabric-Based Home Goods         gosford           au   
1                             Book Stores  st. catharines           ca   
2                             Book Stores  st. catharines           ca   
3       Other Building Material Retailers        yarmouth           ca   
4                 Plastic Surgery Clinics     los angeles           us   
...                                   ...             ...          ...   
232113                                NaN             NaN          NaN   
232114                                NaN             NaN          NaN   
232115                                NaN             NaN          NaN   
232116                                NaN             NaN          NaN   
232117                                NaN             NaN          NaN   

         country_name                                name       phone_x  \
0           australia              Spotlight West Gosford  +61243355946   
1              canada       Heritage Christian Book Store  +19059374553   
2              canada       Heritage Christian Book Store  +19059374553   
3              canada                Pleasant Timber Mart  +19027429181   
4       united states  Skin Specifics Medical Spa West LA  +18184268353   
...               ...                                 ...           ...   
232113            NaN                                 NaN           NaN   
232114            NaN                                 NaN           NaN   
232115            NaN                                 NaN           NaN   
232116            NaN                                 NaN           NaN   
232117            NaN                                 NaN           NaN   

       phone_country_code region_code      region_name  ... domain_suffix  \
0                      au         nsw  new south wales  ...           NaN   
1                      ca          on          ontario  ...           NaN   
2                      ca          on          ontario  ...           NaN   
3                      ca          ns      nova scotia  ...           NaN   
4                      us          ca       california  ...           NaN   
...                   ...         ...              ...  ...           ...   
232113                NaN         NaN              NaN  ...            ca   
232114                NaN         NaN              NaN  ...            ca   
232115                NaN         NaN              NaN  ...           com   
232116                NaN         NaN              NaN  ...            ca   
232117                NaN         NaN              NaN  ...           com   

       language                                         legal_name  \
0           NaN                                                NaN   
1           NaN                                                NaN   
2           NaN                                                NaN   
3           NaN                                                NaN   
4           NaN                                                NaN   
...         ...                                                ...   
232113       en                        Long Point Pest Control Ltd   
232114       fr                           

In [33]:
Datasets_joined.columns.values

array(['address', 'category', 'city', 'country_code', 'country_name',
       'name', 'phone_x', 'phone_country_code', 'region_code',
       'region_name', 'description', 'zip_code', 'domain', 'email',
       'link', 'page_type', 'root_domain', 'domain_suffix', 'language',
       'legal_name', 'main_city', 'main_country', 'main_region',
       'phone_y', 'site_name', 'tld', 's_category'], dtype=object)

Now we do the same thing we did before but we also have to do it for the name column because for the company names unique to the website dataset we will need to copy them to the name column and the legal_name column should be droped.

In [34]:
#Handling name columns
Datasets_joined.loc[Datasets_joined["name"].isnull(), "name"] = Datasets_joined["legal_name"]
del Datasets_joined['legal_name']

In [35]:
#Handling category columns
Datasets_joined.loc[Datasets_joined["category"].isnull(), "category"] = Datasets_joined["s_category"]
del Datasets_joined['s_category']

In [36]:
#Handling city columns
Datasets_joined.loc[Datasets_joined["city"].isnull(), "city"] = Datasets_joined["main_city"]
del Datasets_joined['main_city']

In [37]:
#Handling country_name columns
Datasets_joined.loc[Datasets_joined["country_name"].isnull(), "country_name"] = Datasets_joined["main_country"]
del Datasets_joined['main_country']

In [38]:
#Handling region_name columns
Datasets_joined.loc[Datasets_joined["region_name"].isnull(), "region_name"] = Datasets_joined["main_region"]
del Datasets_joined['main_region']

In [39]:
#Handling phone columns
Datasets_joined.loc[Datasets_joined["phone_x"].isnull(), "phone_x"] = Datasets_joined["phone_y"]
del Datasets_joined['phone_y']
Datasets_joined.rename(columns = {'phone_x':'phone'}, inplace = True)

In [40]:
#Handling domain columns
Datasets_joined.loc[Datasets_joined["domain"].isnull(), "domain"] = Datasets_joined["root_domain"]
del Datasets_joined['root_domain']

Now we can check for the final dataset if our important data has missing values.

In [41]:
print(Datasets_joined.isnull().any())

address                True
category              False
city                  False
country_code           True
country_name          False
name                  False
phone                 False
phone_country_code     True
region_code            True
region_name           False
description            True
zip_code               True
domain                False
email                  True
link                   True
page_type              True
domain_suffix          True
language               True
site_name              True
tld                    True
dtype: bool


As we can see there are null values in our addresses column because the website dataset does not have any address. But our Category, City, Country, Name, Phone, Region and Domain do not have any null values.

In [42]:
display(Datasets_joined)

address  \
0       28 Central Coast Hwy, West Gosford NSW 2250, A...   
1        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
2        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
3           191 Pleasant St, Yarmouth, NS B5A 2J9, Canada   
4       11040 Santa Monica Blvd Suite 370, Los Angeles...   
...                                                   ...   
232113                                                NaN   
232114                                                NaN   
232115                                                NaN   
232116                                                NaN   
232117                                                NaN   

                                           category            city  \
0                           Fabric-Based Home Goods         gosford   
1                                       Book Stores  st. catharines   
2                                       Book Stores  st. catharines   
3                 Other Building Material Retailers        yarmouth   
4                           Plastic Surgery Clinics     los angeles   
...                                             ...             ...   
232113       Pest Services & Agricultural Chemicals  norfolk county   
232114                             Vaporizer Stores     chateauguay   
232115  Investment Consultants & Financial Advisors        oakville   
232116                 Digital & Marketing Agencies           miami   
232117              Decorators & Interior Designers  qualicum beach   

       country_code   country_name  \
0                au      australia   
1                ca         canada   
2                ca         canada   
3                ca         canada   
4                us  united states   
...             ...            ...   
232113          NaN         canada   
232114          NaN         canada   
232115          NaN         canada   
232116          NaN  united states   
232117          NaN         canada   

                                                     name         phone  \
0                                  Spotlight West Gosford  +61243355946   
1                           Heritage Christian Book Store  +19059374553   
2                           Heritage Christian Book Store  +19059374553   
3                                    Pleasant Timber Mart  +19027429181   
4                      Skin Specifics Medical Spa West LA  +18184268353   
...                                                   ...           ...   
232113                        Long Point Pest Control Ltd   15195868585   
232114                                     la Vapote inc.   18333020247   
232115  Connor, Clark & Lunn Private Capital Ltd. (CC&...   18009640520   
232116                     Absolute Data Destruction Inc.   14167427411   
232117                      Studio AE Interior Design Inc   12503902505   

       phone_country_code region_code       region_name  \
0                      au         nsw   new south wales   
1                      ca          on           ontario   
2                      ca          on           ontario   
3                      ca          ns       nova scotia   
4                      us          ca        california   
...                   ...         ...               ...   
232113                NaN         NaN           ontario   
232114                NaN         NaN            quebec   
232115                NaN         NaN           ontario   
232116                NaN         NaN           florida   
232117                NaN         NaN  british columbia   

                                              description zip_code  \
0       4.1 (766) · Craft store West Gosford NSW, Aust...     2250   
1       4.7 (100) · Book store 400 Scott St · In Grant...  l2m 3w2   
2       4.7 (101) · Book store 400 Scott St · In Grant...  l2m 3w2   
3       4.7 (40) · Building materials store 7+ years i...  b5a 2j9   
4       4.3 (15) · Medical spa 7+ years in busi

In [43]:
Datasets_joined.columns.values

array(['address', 'category', 'city', 'country_code', 'country_name',
       'name', 'phone', 'phone_country_code', 'region_code',
       'region_name', 'description', 'zip_code', 'domain', 'email',
       'link', 'page_type', 'domain_suffix', 'language', 'site_name',
       'tld'], dtype=object)

#**Handling duplicates**

For duplicate values I will check 3 times based first on the company name then the adress and at last the domain. I will then delete the duplicate or duplicates.

In [45]:
Datasets_joined.drop_duplicates(subset=['name', 'address', 'domain'], keep='first', inplace=False)

address  \
0       28 Central Coast Hwy, West Gosford NSW 2250, A...   
1        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
2        400 Scott St, St. Catharines, ON L2M 3W2, Canada   
3           191 Pleasant St, Yarmouth, NS B5A 2J9, Canada   
4       11040 Santa Monica Blvd Suite 370, Los Angeles...   
...                                                   ...   
232113                                                NaN   
232114                                                NaN   
232115                                                NaN   
232116                                                NaN   
232117                                                NaN   

                                           category            city  \
0                           Fabric-Based Home Goods         gosford   
1                                       Book Stores  st. catharines   
2                                       Book Stores  st. catharines   
3                 Other Building Material Retailers        yarmouth   
4                           Plastic Surgery Clinics     los angeles   
...                                             ...             ...   
232113       Pest Services & Agricultural Chemicals  norfolk county   
232114                             Vaporizer Stores     chateauguay   
232115  Investment Consultants & Financial Advisors        oakville   
232116                 Digital & Marketing Agencies           miami   
232117              Decorators & Interior Designers  qualicum beach   

       country_code   country_name  \
0                au      australia   
1                ca         canada   
2                ca         canada   
3                ca         canada   
4                us  united states   
...             ...            ...   
232113          NaN         canada   
232114          NaN         canada   
232115          NaN         canada   
232116          NaN  united states   
232117          NaN         canada   

                                                     name         phone  \
0                                  Spotlight West Gosford  +61243355946   
1                           Heritage Christian Book Store  +19059374553   
2                           Heritage Christian Book Store  +19059374553   
3                                    Pleasant Timber Mart  +19027429181   
4                      Skin Specifics Medical Spa West LA  +18184268353   
...                                                   ...           ...   
232113                        Long Point Pest Control Ltd   15195868585   
232114                                     la Vapote inc.   18333020247   
232115  Connor, Clark & Lunn Private Capital Ltd. (CC&...   18009640520   
232116                     Absolute Data Destruction Inc.   14167427411   
232117                      Studio AE Interior Design Inc   12503902505   

       phone_country_code region_code       region_name  \
0                      au         nsw   new south wales   
1                      ca          on           ontario   
2                      ca          on           ontario   
3                      ca          ns       nova scotia   
4                      us          ca        california   
...                   ...         ...               ...   
232113                NaN         NaN           ontario   
232114                NaN         NaN            quebec   
232115                NaN         NaN           ontario   
232116                NaN         NaN           florida   
232117                NaN         NaN  british columbia   

                                              description zip_code  \
0       4.1 (766) · Craft store West Gosford NSW, Aust...     2250   
1       4.7 (100) · Book store 400 Scott St · In Grant...  l2m 3w2   
2       4.7 (101) · Book store 400 Scott St · In Grant...  l2m 3w2   
3       4.7 (40) · Building materials store 7+ years i...  b5a 2j9   
4       4.3 (15) · Medical spa 7+ years in busi

Looks like we had only 112 duplicates.

#**Saving the final dataset to csv**

In [49]:
Datasets_joined.to_csv("/content/Datasets_joined.csv", index=False)